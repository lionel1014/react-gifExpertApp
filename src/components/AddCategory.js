import React, { useState } from 'react'
import PropTypes from 'prop-types'


export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');  // => ('') > () => ('') != () => ( ) da una error, envia una variale undefined

    const handleInputChange = (e) =>{//los valores que recibimos del input
        const {value} = e.target;
        setInputValue ( value );

        // console.log('handleInputChange fue llamado con exito')

    };

    const handleSubmit = (e) =>{//para enviar el input en un form
        e.preventDefault();
        
        if ( inputValue.trim().length >= 3 ){
            setCategories(cate => [inputValue,...cate]); // guardamos el nuevo valor del `inputValue`
            setInputValue (''); // el estado del inputValue ahora es vacío
        }

    }

    return (
        <form onSubmit = {handleSubmit} > 
            <p> {inputValue} </p>
            <input type="text" value={inputValue} placeholder='Ingresa una categoría'
            onChange = { handleInputChange }
            />
        </form>
    )
}

AddCategory.propTypes = {

    setCategories : PropTypes.func.isRequired

};
