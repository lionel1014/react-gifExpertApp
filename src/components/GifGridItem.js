import React from 'react';
import PropTypes from 'prop-types'; 

export const GifGridItem = ({title,url}) => {

    // console.log({title,url})

    return (
        <div className="imagenes animate__animated animate__flipInX" >
            <p> {title} </p>
            <img src={ url } alt= {title} />
        </div>
    )
}

GifGridItem.propTypes ={

    title : PropTypes.string.isRequired,
    url: PropTypes.string.isRequired
};
