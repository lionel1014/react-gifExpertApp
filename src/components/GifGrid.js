import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => { // no te olvides de desestructurar la propiedad

    const {data:images,loading} = useFetchGifs(category);  //data:images, en la desestructuracion cambiamos el nombre de la variable asi
    
    // useEffect(() =>{
    //     //importamos la promesa getGifs de "helpers"
    //     getGifs(category)
    //         .then(setImages)
    //         // .then( imgs => setImages(imgs)) // una funcion que manda el primer argumento al primer argumento de setImages
    //         // se lo puede resumir como el de arriba
    // },[category]); // que haga lo que hay dentro si cambia category

    return (
        <>
        <h3 className="animate__animated animate__bounceInLeft "> {category} </h3>
        {/* {loading ? 'Cargando...' : 'Fin de la carga del objeto.'} */}
         {loading && <p className="animate__animated animate__bounceIn " >Cargando...</p>}  {/*el && es una condición lógica AND */}
        
        <div className="imagenesGrid" >
            
            {
                images.map( ( img ) =>{
                    return <GifGridItem 
                            key = {img.id}
                            {...img} 
                            />
                }) 
            }
            
        </div>
        </>
    )
}
