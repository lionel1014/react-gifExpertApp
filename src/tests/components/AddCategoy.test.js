import '@testing-library/jest-dom';
import { shallow } from 'enzyme';
import {AddCategory} from '../../components/AddCategory'

describe('Pruebas en AddCategoy', () => {

const setCategories = jest.fn(); //funcion de prueba
let wrapper = shallow( <AddCategory setCategories={setCategories} /> ); // hay que enviarle algo ya que es requerido
    
    beforeEach(() =>{
        jest.clearAllMocks(); //sirve para limpiar las simulaciones
        wrapper = shallow( <AddCategory setCategories={setCategories} /> );
    });

    test('Debe de mostrarse correctamente (Snapshot) ', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('Debe cambiar la caja de texto ', () => {
        
        const input = wrapper.find('input');
        const value = 'Probando la caja de texto'

        input.simulate('change', {target: {value} }); // el evento de change se define asi, 

        expect(wrapper.find('p').text().trim()).toBe(value) //aqui esperamos que guarde la capture del evento de la caja de texto
    });
    
    test('NO debe enviar el SUBMIT', () => {
        
        wrapper.find('form').simulate('submit',{
            preventDefault(){

            } });  // simulados el envio del form

        expect( setCategories ).not.toHaveBeenCalled();

    });
    
    test('Debe llamar a setCategories y limpiar la caja de texto del input', () => {
        
        // 1) simular el inputChange        
        const input = wrapper.find('input');
        const value = 'Simulando inputChange';

        input.simulate('change', {target: {value} }); // simulamos el cambio del inputChange

        // 2) simular el submit
        wrapper.find('form').simulate('submit',{preventDefault(){} }); // simulamos el envio del form

        // 3) setCategories debe habersellamado
        expect( setCategories ).toHaveBeenCalled();
        expect( setCategories ).toHaveBeenCalledTimes(1); // pregunta si llamo a la funcion X veces
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) ); //esperamos que al menos llame una funci√≥n
        
        // 4) el valor del input debe de estar ''
        expect(wrapper.find('input').prop('value')).toBe('');
        //abajo como yo lo hice
        // input.simulate('change', {target: {} }); //seteo en vacio el input
        // expect(wrapper.find('p').text().trim()).toBe(''); //pregunto si esta vacio

    })
    

})
