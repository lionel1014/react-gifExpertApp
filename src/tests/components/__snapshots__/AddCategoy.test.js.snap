// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AddCategoy Debe de mostrarse correctamente (Snapshot)  1`] = `
<form
  onSubmit={[Function]}
>
  <p>
     
     
  </p>
  <input
    onChange={[Function]}
    placeholder="Ingresa una categorÃ­a"
    type="text"
    value=""
  />
</form>
`;
